File: src/verdicts.ts
import { existsSync } from "fs";
import { join } from "path";
import { fail } from "./errors";
import { appendEvent, ensureDir, nowIso, readJson, readTasks, writeJson } from "./io";
import { CRITIC_VERDICTS_DIR } from "./paths";
import {
  CheckStatus,
  CriticVerdict,
  CriticVerdictSchema,
  VerdictCheck,
  VerdictDecision,
} from "./schemas";

export function writeCriticVerdict(
  root: string,
  taskId: string,
  verdict: VerdictDecision,
  summary: string,
  checkArgs: string[],
  gaps: string[],
  actor: string,
): void {
  ensureTaskExists(root, taskId);
  ensureDir(join(root, CRITIC_VERDICTS_DIR));

  const createdAt = nowIso();
  const verdictDoc: CriticVerdict = {
    schemaVersion: 1,
    taskId,
    verdict,
    summary,
    checks: checkArgs.map(parseVerdictCheckArg),
    gaps,
    createdAt,
  };

  const safeTaskId = safeFileId(taskId);
  const timestamp = compactTimestamp(createdAt);
  const verdictPath = join(root, CRITIC_VERDICTS_DIR, `${safeTaskId}-${timestamp}.json`);
  const latestPath = latestVerdictPath(root, taskId);
  writeJson(verdictPath, CriticVerdictSchema, verdictDoc);
  writeJson(latestPath, CriticVerdictSchema, verdictDoc);

  appendEvent(root, {
    timestamp: nowIso(),
    kind: "critic.verdict_written",
    actor,
    taskId,
    message: `Critic verdict for ${taskId}: ${verdict}`,
    data: { verdict, latestPath: latestVerdictRelativePath(taskId) },
  });

  console.log(`wrote ${latestPath}`);
}

export function ensureLatestVerdictApproved(root: string, taskId: string): void {
  const verdict = readLatestVerdict(root, taskId);
  if (verdict.taskId !== taskId) {
    fail(`latest critic verdict task mismatch: expected ${taskId}, got ${verdict.taskId}`);
  }
  if (verdict.verdict !== "approved") {
    fail(`latest critic verdict for task ${taskId} is ${verdict.verdict}, not approved`);
  }
}

export function readLatestVerdict(root: string, taskId: string): CriticVerdict {
  const path = latestVerdictPath(root, taskId);
  if (!existsSync(path)) {
    fail(`task ${taskId} cannot complete without an approved critic verdict: read ${path}`);
  }
  return readJson(path, CriticVerdictSchema, path);
}

export function parseVerdictDecision(raw: string): VerdictDecision {
  const normalized = raw.trim().toLowerCase().replace(/-/g, "_");
  const allowed: VerdictDecision[] = ["approved", "rejected", "human_approval_required"];
  if (!allowed.includes(normalized as VerdictDecision)) {
    fail(`unknown verdict ${normalized}`);
  }
  return normalized as VerdictDecision;
}

export function parseVerdictCheckArg(raw: string): VerdictCheck {
  const equals = raw.indexOf("=");
  if (equals < 0) {
    fail(`invalid check ${JSON.stringify(raw)}; expected name=status:evidence`);
  }
  const colon = raw.indexOf(":", equals + 1);
  if (colon < 0) {
    fail(`invalid check ${JSON.stringify(raw)}; expected name=status:evidence`);
  }
  return {
    name: raw.slice(0, equals).trim(),
    status: parseCheckStatus(raw.slice(equals + 1, colon)),
    evidence: raw.slice(colon + 1).trim(),
  };
}

export function latestVerdictPath(root: string, taskId: string): string {
  return join(root, latestVerdictRelativePath(taskId));
}

function latestVerdictRelativePath(taskId: string): string {
  return `${CRITIC_VERDICTS_DIR}/${safeFileId(taskId)}.latest.json`;
}

function ensureTaskExists(root: string, taskId: string): void {
  const ledger = readTasks(root);
  if (!ledger.tasks.some((task) => task.id === taskId)) {
    fail(`task ${taskId} not found`);
  }
}

function parseCheckStatus(raw: string): CheckStatus {
  const normalized = raw.trim().toLowerCase().replace(/-/g, "_");
  const allowed: CheckStatus[] = ["passed", "failed", "skipped"];
  if (!allowed.includes(normalized as CheckStatus)) {
    fail(`unknown check status ${normalized}`);
  }
  return normalized as CheckStatus;
}

function safeFileId(value: string): string {
  return value.replace(/[^A-Za-z0-9_-]/g, "-");
}

function compactTimestamp(iso: string): string {
  return iso.replace(/[-:]/g, "").replace(".", "");
}

File: src/verdicts.rs
use crate::{
    fsio::{append_event, ensure_dir, pretty_json},
    models::{CheckStatus, CriticVerdict, Event, VerdictCheck, VerdictDecision},
    paths::CRITIC_VERDICTS_DIR,
    tasks,
};
use anyhow::{bail, Context, Result};
use chrono::Utc;
use serde_json::json;
use std::{collections::BTreeMap, fs, path::Path};

pub(crate) fn write_critic_verdict(
    root: &Path,
    task_id: &str,
    verdict: VerdictDecision,
    summary: &str,
    check_args: &[String],
    gaps: Vec<String>,
    actor: &str,
) -> Result<()> {
    tasks::ensure_task_exists(root, task_id)?;
    ensure_dir(root.join(CRITIC_VERDICTS_DIR))?;

    let checks = check_args
        .iter()
        .map(|arg| parse_verdict_check_arg(arg))
        .collect::<Result<Vec<_>>>()?;

    let verdict_doc = CriticVerdict {
        schema_version: 1,
        task_id: task_id.to_string(),
        verdict,
        summary: summary.to_string(),
        checks,
        gaps,
        created_at: Utc::now(),
    };

    let safe_task_id = safe_file_id(task_id);
    let timestamp = verdict_doc.created_at.format("%Y%m%dT%H%M%SZ");
    let verdict_path = root
        .join(CRITIC_VERDICTS_DIR)
        .join(format!("{safe_task_id}-{timestamp}.json"));
    let latest_path = latest_verdict_path(root, task_id);
    let payload = pretty_json(&verdict_doc)?;
    fs::write(&verdict_path, &payload)
        .with_context(|| format!("write {}", verdict_path.display()))?;
    fs::write(&latest_path, payload).with_context(|| format!("write {}", latest_path.display()))?;

    let mut data = BTreeMap::new();
    data.insert("verdict".to_string(), json!(verdict.as_str()));
    data.insert(
        "latestPath".to_string(),
        json!(latest_verdict_relative_path(task_id)),
    );
    append_event(
        root,
        Event {
            timestamp: Utc::now(),
            kind: "critic.verdict_written".to_string(),
            actor: actor.to_string(),
            task_id: Some(task_id.to_string()),
            message: format!("Critic verdict for {task_id}: {}", verdict.as_str()),
            data,
        },
    )?;

    println!("wrote {}", latest_path.display());
    Ok(())
}

pub(crate) fn ensure_latest_verdict_approved(root: &Path, task_id: &str) -> Result<()> {
    let verdict = read_latest_verdict(root, task_id).with_context(|| {
        format!("task {task_id} cannot complete without an approved critic verdict")
    })?;
    if verdict.task_id != task_id {
        bail!(
            "latest critic verdict task mismatch: expected {task_id}, got {}",
            verdict.task_id
        );
    }
    if verdict.verdict != VerdictDecision::Approved {
        bail!(
            "latest critic verdict for task {task_id} is {}, not approved",
            verdict.verdict.as_str()
        );
    }
    Ok(())
}

fn read_latest_verdict(root: &Path, task_id: &str) -> Result<CriticVerdict> {
    let path = latest_verdict_path(root, task_id);
    let raw = fs::read_to_string(&path).with_context(|| format!("read {}", path.display()))?;
    serde_json::from_str(&raw).with_context(|| format!("parse {}", path.display()))
}

fn latest_verdict_path(root: &Path, task_id: &str) -> std::path::PathBuf {
    root.join(latest_verdict_relative_path(task_id))
}

fn latest_verdict_relative_path(task_id: &str) -> String {
    format!(
        "{}/{}.latest.json",
        CRITIC_VERDICTS_DIR,
        safe_file_id(task_id)
    )
}

fn parse_verdict_check_arg(raw: &str) -> Result<VerdictCheck> {
    let (name, rest) = raw
        .split_once('=')
        .with_context(|| format!("invalid check {raw:?}; expected name=status:evidence"))?;
    let (status, evidence) = rest
        .split_once(':')
        .with_context(|| format!("invalid check {raw:?}; expected name=status:evidence"))?;
    Ok(VerdictCheck {
        name: name.trim().to_string(),
        status: parse_check_status(status)?,
        evidence: evidence.trim().to_string(),
    })
}

fn parse_check_status(status: &str) -> Result<CheckStatus> {
    match status
        .trim()
        .to_ascii_lowercase()
        .replace('-', "_")
        .as_str()
    {
        "passed" => Ok(CheckStatus::Passed),
        "failed" => Ok(CheckStatus::Failed),
        "skipped" => Ok(CheckStatus::Skipped),
        other => bail!("unknown check status {other}"),
    }
}

fn safe_file_id(value: &str) -> String {
    value
        .chars()
        .map(|ch| {
            if ch.is_ascii_alphanumeric() || ch == '-' || ch == '_' {
                ch
            } else {
                '-'
            }
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_verdict_check_arg() {
        let check = parse_verdict_check_arg("cargo check=passed:clean").unwrap();
        assert_eq!(check.name, "cargo check");
        assert_eq!(check.status, CheckStatus::Passed);
        assert_eq!(check.evidence, "clean");
    }

    #[test]
    fn completed_status_requires_approved_verdict() {
        let root = std::env::temp_dir().join(format!(
            "yoloop-verdict-test-{}-{}",
            std::process::id(),
            Utc::now().timestamp_nanos_opt().unwrap()
        ));
        fs::create_dir_all(root.join(CRITIC_VERDICTS_DIR)).unwrap();

        assert!(ensure_latest_verdict_approved(&root, "T-1").is_err());

        let rejected = CriticVerdict {
            schema_version: 1,
            task_id: "T-1".to_string(),
            verdict: VerdictDecision::Rejected,
            summary: "not yet".to_string(),
            checks: vec![],
            gaps: vec![],
            created_at: Utc::now(),
        };
        fs::write(
            latest_verdict_path(&root, "T-1"),
            serde_json::to_string_pretty(&rejected).unwrap(),
        )
        .unwrap();
        assert!(ensure_latest_verdict_approved(&root, "T-1").is_err());

        let approved = CriticVerdict {
            verdict: VerdictDecision::Approved,
            ..rejected
        };
        fs::write(
            latest_verdict_path(&root, "T-1"),
            serde_json::to_string_pretty(&approved).unwrap(),
        )
        .unwrap();
        assert!(ensure_latest_verdict_approved(&root, "T-1").is_ok());

        let _ = fs::remove_dir_all(root);
    }
}
